### PR TITLE
refactor associated type values, introduce `ClauseBuilder`

### PR DESCRIPTION
**Pull request by nikomatsakis**
The main goal of this PR (which builds on #265) is to introduce the `ClauseBuilder` API, which aims to make the `program_clauses` code rather more readable. In particular, we no longer manually shift de bruijn indices hither and thither, but rather just rely on substitution. 

The core `ClauseBuilder` API are three methods:

* `push_binders`, which lets you add binders into scope;
* `push_clause`, which pushes a program clause (universally quantified over all binders that are in scope);
* `push_fact`, which is just a `push_clause` with no conditions.

So for example given a `Binders<TraitRef>`, like `forall<T> { T: Foo }`, we could make a clause:

```
forall<T> {
   Implemented(T: Foo).
}
```

via code like

```rust
let binders: Binders<TraitRef> = ...;
builder.push_binders(&binders, |builder, trait_ref| {
    // `trait_ref` is the trait reference `T: Foo`
    builder.push_fact(DomainGoal::Implemented(trait_ref));
});
```

r? @tmandry 
cc @scalexm

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Rebased.

File: chalk-solve/src/clauses.rs [hunk 1]
```diff
@@ -45,36 +46,39 @@ pub mod program_clauses;
...
     // for Foo<..>`, where `Foo` is the struct we're looking at, then
     // we don't generate our own rules.
-    if program.impl_provided_for(auto_trait_id, struct_id) {
+    if builder.db.impl_provided_for(auto_trait_id, struct_id) {
         debug!("impl provided");
         return;
     }
 
-    vec.push({
+    let binders = struct_datum.binders.map_ref(|b| (&b.self_ty, &b.fields));
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/clauses.rs]: Just a micro-optimization of sorts; it avoids folding the parts we don't actually use.

File: chalk-solve/src/clauses/program_clauses.rs [hunk 2]
```diff
@@ -676,172 +676,126 @@ impl ToProgramClauses for AssociatedTyDatum {
...
+                parameters: parameters.clone(),
+            };
 
-        let projection_eq = ProjectionEq {
-            projection: projection.clone(),
-            ty: app_ty.clone(),
-        };
+            // Retrieve the trait ref embedding the associated type
+            let trait_ref = {
+                let (associated_ty_data, trait_params, _) = db.split_projection(&projection);
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/clauses/program_clauses.rs]: Answer: given a GAT projection like `(Iteratable::Item)<T, 'a>`, this splits the substitutions into those for the trait (`T: Iterable`) and those for the `Item` (`'a`).

File: chalk-solve/src/clauses/program_clauses.rs [hunk 3]
```diff
@@ -447,9 +447,20 @@ impl ToProgramClauses for TraitDatum {
...
+        let parameters: Vec<_> = self
+            .binders
+            .binders
+            .iter()
+            .zip(0..)
+            .map(|p| p.to_parameter())
+            .collect();
 
-        let trait_ref_impl = WhereClause::Implemented(self.binders.value.trait_ref.clone());
+        let trait_ref = chalk_ir::TraitRef {
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/clauses/program_clauses.rs]: Do you just mean introducing `TraitRef::new`? I'd have no objection...

[the remark this answers is not in the document]
Comment (tmandry) [reply chalk-solve/src/clauses/program_clauses.rs]: ah no, sorry, I specifically meant building the list of parameters.

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/clauses/program_clauses.rs]: that does get fixed in later commits

File: chalk-solve/src/clauses/program_clauses.rs [hunk 4]
```diff
@@ -82,65 +82,54 @@ impl ToProgramClauses for AssociatedTyValue {
...
+            let all_parameters = builder.placeholders_in_scope().to_vec();
 
-        let normalize_goal = DomainGoal::Normalize(Normalize {
-            projection,
-            ty: self.value.value.ty.clone(),
-        });
+            // Get the projection for this associated type:
+            //
+            // * `impl_params`: `[!T]`
+            // * `projection`: `<Vec<!T> as Iterable>::Iter<'!a`>
```
Comment (tmandry) [inline chalk-solve/src/clauses/program_clauses.rs]: Nit: misplaced backtick

File: chalk-solve/src/clauses/program_clauses.rs [hunk 5]
```diff
@@ -408,167 +407,119 @@ impl ToProgramClauses for TraitDatum {
...
+                        .chain(iter::once(Goal::CannotProve(()))),
+                );
+            }
+
+            // Orphan rules:
+            if !self.flags.upstream {
+                // Impls for traits declared locally always pass the impl rules
+                builder.push_fact(DomainGoal::LocalImplAllowed(trait_ref.clone()));
+            } else {
+                // Impls for remote traits must have a local type int he right place
```
Comment (tmandry) [inline chalk-solve/src/clauses/program_clauses.rs]: Nit: in the

File: chalk-solve/src/clauses.rs [hunk 6]
```diff
@@ -169,10 +169,31 @@ fn program_clauses_that_could_match(
...
+            // })
+            // ```
+            //
+            // so what we wish to do is to generate program clauses of the form:
+            //
+            // ```
+            // forall<'a> { Implemented(dyn Fn(&u8): Fn<(&'a u8)>) }
+            // ```
+            //
+            // or
```
Comment (tmandry) [inline chalk-solve/src/clauses.rs]: nit: this “or” tripped me up at first, maybe add something about “, whatever the specified goal is trying to prove” (or don’t, it’s sorta obvious from context)

File: chalk-solve/src/wf.rs [hunk 7]
```diff
@@ -351,4 +284,91 @@ where
...
+    /// trait Foo {
+    ///     type Item<'a>
+    /// }
+    ///
+    /// impl<T> Foo for Box<T> {
+    ///     type Item<'a> = Box<&'a T>;
+    /// }
+    /// ```
+    ///
+    /// we would issue the following subgoal: `forall<'a> { WellFormed(Box<&'a T>) }`.
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/wf.rs]: I'm not sure quite what you have in mind, but I've expanded the comments some more.

File: chalk-solve/src/split.rs [hunk 8]
```diff
@@ -97,6 +97,61 @@ pub trait Split: RustIrDatabase {
...
+    ///
+    /// Here we expect the full set of parameters for `Iter`, which
+    /// would be `['x, Y]`, where `'x` is the value for `'a` and `Y`
+    /// is the value for `T`.
+    ///
+    /// Returns the pair of:
+    ///
+    /// * the trait-ref `Vec<Y>: Iterable`
+    /// * the projection `<Vec<Y> as Iterable>::Iter<'x>`
+    fn impl_trait_ref_and_projection_from_associated_ty_value(
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/split.rs]: It seems like I removed it in later commits, so I guess maybe I agreed with you =) 

We now have only one routine:

```rust
    fn impl_parameters_and_projection_from_associated_ty_value<'p>(
        &self,
        parameters: &'p [Parameter<ChalkIr>],
        associated_ty_value: &AssociatedTyValue,
    ) -> (&'p [Parameter<ChalkIr>], ProjectionTy<ChalkIr>) {
 ```

It seems reasonably clear what it does.

File: chalk-rust-ir/src/lib.rs [hunk 9]
```diff
@@ -321,6 +321,10 @@ pub struct AssociatedTyDatumBound {
     pub where_clauses: Vec<QuantifiedWhereClause<ChalkIr>>,
 }
 
+impl HasTypeFamily for AssociatedTyDatumBound {
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-rust-ir/src/lib.rs]: Yes -- that is orthogonal to this PR but should I think be improved. That annotation only guides the `Fold` derive. But I think it should also generate this impl. We should file an issue.

File: chalk-solve/src/wf.rs [hunk 10]
```diff
@@ -333,23 +319,40 @@ where
...
-        let bound_goals = assoc_ty_datum_bounds
-            .into_iter()
-            .map(|quantified_bound| quantified_bound.substitute(&projection.parameters))
-            .flat_map(|b| b.into_where_clauses(assoc_ty.value.value.ty.clone()))
+        // Get the bounds and where clauses from the trait
+        // declaration, substituted appropriately.
+        //
+        // From our example:
+        //
+        // * bounds - in trait, `Clone`, in impl, `Clone`
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/wf.rs]: I think the comment was poorly rephrased. I've changed it:

```rust
        // Get the bounds and where clauses from the trait
        // declaration, substituted appropriately.
        //
        // From our example:
        //
        // * bounds
        //     * original in trait, `Clone`
        //     * after substituting impl parameters, `Clone`
        // * where clauses
        //     * original in trait, `Self: 'a`
        //     * after substituting impl parameters, `Box<!T>: '!a`
```

File: chalk-solve/src/wf.rs [hunk 11]
```diff
@@ -333,23 +319,40 @@ where
...
+        // * where clauses - in trait, `Self: 'a`, in impl, `Box<!T>: '!a`
+        let assoc_ty_datum = self.db.associated_ty_data(projection.associated_ty_id);
+        let AssociatedTyDatumBound {
+            bounds: defn_bounds,
+            where_clauses: defn_where_clauses,
+        } = assoc_ty_datum.binders.substitute(&projection.parameters);
+
+        // Check that the `value_ty` meets the bounds (i.e., `Box<!T>: Clone`).
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply chalk-solve/src/wf.rs]: reworded and expanded

File: chalk-solve/src/clauses.rs (hunk 6 above)
Comment (nikomatsakis) [reply chalk-solve/src/clauses.rs]: improved

**Pull request merged**